File: extensions/rust/tests/test_neighbors.py
"""Tests for the O(n) cell-list neighbor finding implementation."""

import json
import time

import numpy as np
import pytest

import ferrox


def make_fcc_cu(lattice_const: float = 3.61) -> dict:
    """Create FCC Cu structure."""
    return {
        "@module": "pymatgen.core.structure",
        "@class": "Structure",
        "lattice": {
            "matrix": [
                [lattice_const, 0.0, 0.0],
                [0.0, lattice_const, 0.0],
                [0.0, 0.0, lattice_const],
            ],
            "pbc": [True, True, True],
        },
        "sites": [
            {"species": [{"element": "Cu", "occu": 1.0}], "abc": [0.0, 0.0, 0.0]},
            {"species": [{"element": "Cu", "occu": 1.0}], "abc": [0.5, 0.5, 0.0]},
            {"species": [{"element": "Cu", "occu": 1.0}], "abc": [0.5, 0.0, 0.5]},
            {"species": [{"element": "Cu", "occu": 1.0}], "abc": [0.0, 0.5, 0.5]},
        ],
    }


def make_bcc_fe(lattice_const: float = 2.87) -> dict:
    """Create BCC Fe structure."""
    return {
        "@module": "pymatgen.core.structure",
        "@class": "Structure",
        "lattice": {
            "matrix": [
                [lattice_const, 0.0, 0.0],
                [0.0, lattice_const, 0.0],
                [0.0, 0.0, lattice_const],
            ],
            "pbc": [True, True, True],
        },
        "sites": [
            {"species": [{"element": "Fe", "occu": 1.0}], "abc": [0.0, 0.0, 0.0]},
            {"species": [{"element": "Fe", "occu": 1.0}], "abc": [0.5, 0.5, 0.5]},
        ],
    }


def make_nacl(lattice_const: float = 5.64) -> dict:
    """Create NaCl rocksalt structure."""
    return {
        "@module": "pymatgen.core.structure",
        "@class": "Structure",
        "lattice": {
            "matrix": [
                [lattice_const, 0.0, 0.0],
                [0.0, lattice_const, 0.0],
                [0.0, 0.0, lattice_const],
            ],
            "pbc": [True, True, True],
        },
        "sites": [
            {"species": [{"element": "Na", "occu": 1.0}], "abc": [0.0, 0.0, 0.0]},
            {"species": [{"element": "Na", "occu": 1.0}], "abc": [0.5, 0.5, 0.0]},
            {"species": [{"element": "Na", "occu": 1.0}], "abc": [0.5, 0.0, 0.5]},
            {"species": [{"element": "Na", "occu": 1.0}], "abc": [0.0, 0.5, 0.5]},
            {"species": [{"element": "Cl", "occu": 1.0}], "abc": [0.5, 0.5, 0.5]},
            {"species": [{"element": "Cl", "occu": 1.0}], "abc": [0.0, 0.0, 0.5]},
            {"species": [{"element": "Cl", "occu": 1.0}], "abc": [0.0, 0.5, 0.0]},
            {"species": [{"element": "Cl", "occu": 1.0}], "abc": [0.5, 0.0, 0.0]},
        ],
    }


def make_large_supercell(n: int = 3) -> dict:
    """Create a large FCC Cu supercell for performance testing."""
    base = make_fcc_cu()
    base_lattice = np.array(base["lattice"]["matrix"])
    base_sites = base["sites"]

    # Create supercell
    new_lattice = (base_lattice * n).tolist()
    new_sites = []

    for idx_a in range(n):
        for idx_b in range(n):
            for idx_c in range(n):
                for site in base_sites:
                    abc = site["abc"]
                    new_abc = [
                        (abc[0] + idx_a) / n,
                        (abc[1] + idx_b) / n,
                        (abc[2] + idx_c) / n,
                    ]
                    new_sites.append({
                        "species": site["species"],
                        "abc": new_abc,
                    })

    return {
        "@module": "pymatgen.core.structure",
        "@class": "Structure",
        "lattice": {
            "matrix": new_lattice,
            "pbc": [True, True, True],
        },
        "sites": new_sites,
    }


class TestNeighborListBasics:
    """Basic neighbor list functionality tests."""

    def test_fcc_coordination_12(self) -> None:
        """FCC structure should have CN=12 for each site."""
        fcc = json.dumps(make_fcc_cu())
        cutoff = 3.0  # covers first shell at ~2.55 Å

        centers, neighbors, images, distances = ferrox.get_neighbor_list(fcc, cutoff)

        # Count neighbors per site
        counts = [0] * 4
        for center in centers:
            counts[center] += 1

        # Each site should have 12 neighbors
        assert all(count == 12 for count in counts), f"Expected CN=12, got {counts}"

    def test_bcc_coordination_8(self) -> None:
        """BCC structure should have CN=8 for first shell."""
        bcc = json.dumps(make_bcc_fe())
        cutoff = 2.6  # covers first shell at ~2.48 Å

        centers, neighbors, images, distances = ferrox.get_neighbor_list(bcc, cutoff)

        # Count neighbors per site
        counts = [0] * 2
        for center in centers:
            counts[center] += 1

        # Each site should have 8 neighbors
        assert all(count == 8 for count in counts), f"Expected CN=8, got {counts}"

    def test_nacl_coordination_6(self) -> None:
        """NaCl structure should have CN=6 for each site."""
        nacl = json.dumps(make_nacl())
        cutoff = 3.5  # covers first shell at ~2.82 Å

        centers, neighbors, images, distances = ferrox.get_neighbor_list(nacl, cutoff)

        # Count neighbors per site
        counts = [0] * 8
        for center in centers:
            counts[center] += 1

        # Each site should have 6 neighbors
        assert all(count == 6 for count in counts), f"Expected CN=6, got {counts}"

    def test_distance_values(self) -> None:
        """Check that distance values are correct for FCC."""
        fcc = json.dumps(make_fcc_cu(3.61))
        cutoff = 3.0

        centers, neighbors, images, distances = ferrox.get_neighbor_list(fcc, cutoff)

        # Expected first shell distance: a/sqrt(2) ≈ 2.552 Å
        expected_dist = 3.61 / np.sqrt(2)

        for dist in distances:
            assert abs(dist - expected_dist) < 0.1, f"Distance {dist} doesn't match expected {expected_dist}"

    def test_empty_result_for_small_cutoff(self) -> None:
        """Very small cutoff should return empty neighbor list."""
        fcc = json.dumps(make_fcc_cu())
        cutoff = 0.1  # too small to find any neighbors

        centers, neighbors, images, distances = ferrox.get_neighbor_list(fcc, cutoff)

        assert len(centers) == 0

    def test_negative_cutoff(self) -> None:
        """Negative cutoff should raise an error."""
        fcc = json.dumps(make_fcc_cu())

        with pytest.raises(ValueError):
            ferrox.get_neighbor_list(fcc, -1.0)


class TestPeriodicImages:
    """Tests for periodic boundary condition handling."""

    def test_periodic_images_present(self) -> None:
        """Neighbor list should include periodic images."""
        fcc = json.dumps(make_fcc_cu())
        cutoff = 3.0

        centers, neighbors, images, distances = ferrox.get_neighbor_list(fcc, cutoff)

        # Some images should be non-zero
        nonzero_images = [img for img in images if img != [0, 0, 0]]
        assert len(nonzero_images) > 0, "Expected some periodic images"

    def test_image_offsets_correct(self) -> None:
        """Image offsets should be small integers."""
        fcc = json.dumps(make_fcc_cu())
        cutoff = 3.0

        centers, neighbors, images, distances = ferrox.get_neighbor_list(fcc, cutoff)

        for img in images:
            assert all(abs(x) <= 1 for x in img), f"Unexpected large image offset: {img}"


class TestPerformance:
    """Performance tests to verify O(n) scaling."""

    @pytest.mark.parametrize("supercell_size", [2, 3, 4])
    def test_scaling_with_system_size(self, supercell_size: int) -> None:
        """Verify that neighbor finding completes in reasonable time."""
        struct = make_large_supercell(supercell_size)
        num_atoms = len(struct["sites"])
        struct_json = json.dumps(struct)
        cutoff = 3.0

        start = time.perf_counter()
        centers, neighbors, images, distances = ferrox.get_neighbor_list(struct_json, cutoff)
        elapsed = time.perf_counter() - start

        # Should complete in well under 1 second even for ~500 atoms
        assert elapsed < 2.0, f"Neighbor finding took {elapsed:.2f}s for {num_atoms} atoms"

        # Verify we found neighbors
        assert len(centers) > 0, f"No neighbors found for {num_atoms} atom system"


class TestConsistency:
    """Tests for consistency with coordination analysis functions."""

    def test_consistency_with_coordination_numbers(self) -> None:
        """Neighbor list should be consistent with coordination numbers."""
        fcc = json.dumps(make_fcc_cu())
        cutoff = 3.0

        # Get neighbor list
        centers, neighbors, images, distances = ferrox.get_neighbor_list(fcc, cutoff)

        # Get coordination numbers
        cns = ferrox.get_coordination_numbers(fcc, cutoff)

        # Count from neighbor list
        nl_counts = [0] * len(cns)
        for center in centers:
            nl_counts[center] += 1

        assert nl_counts == cns, f"Mismatch: NL counts {nl_counts} vs CN {cns}"

    def test_consistency_with_local_environment(self) -> None:
        """Neighbor list should be consistent with local environment."""
        fcc = json.dumps(make_fcc_cu())
        cutoff = 3.0

        # Get neighbor list
        centers, neighbors, images, distances = ferrox.get_neighbor_list(fcc, cutoff)

        # Get local environment for site 0
        local_env = ferrox.get_local_environment(fcc, 0, cutoff)

        # Count neighbors of site 0 from neighbor list
        nl_count = sum(1 for center in centers if center == 0)

        assert nl_count == len(local_env), f"NL count {nl_count} vs local_env {len(local_env)}"

File: extensions/rust/tests/test_neighbors.rs
//! Integration tests for the neighbors module.
//!
//! These tests verify large-scale behavior and cross-cutoff consistency.

mod common;

use common::{make_bcc, make_bcc_large, make_fcc, make_fcc_large, make_nacl, make_nacl_supercell};
use ferrox::element::Element;
use ferrox::neighbors::{NeighborListConfig, build_neighbor_list};
use std::time::Instant;

// === Large System Scaling Tests ===

#[test]
fn test_large_fcc_scaling_1000_atoms() {
    // 6x6x6 FCC supercell = 864 atoms
    let large_fcc = make_fcc_large(3.61, Element::Cu, 6);
    let n_atoms = large_fcc.num_sites();
    assert!(
        n_atoms >= 800,
        "Expected at least 800 atoms, got {}",
        n_atoms
    );

    let config = NeighborListConfig {
        cutoff: 3.0,
        ..Default::default()
    };

    let start = Instant::now();
    let nl = build_neighbor_list(&large_fcc, &config);
    let elapsed = start.elapsed();

    // Performance: should complete in < 2 seconds
    assert!(
        elapsed.as_secs_f64() < 2.0,
        "Took {:.2}s for {} atoms - expected < 2s",
        elapsed.as_secs_f64(),
        n_atoms
    );

    // Correctness: each atom should have 12 neighbors
    let mut counts = vec![0usize; n_atoms];
    for &center in &nl.center_indices {
        counts[center] += 1;
    }

    let all_12 = counts.iter().all(|&c| c == 12);
    assert!(all_12, "All FCC atoms should have CN=12");

    println!(
        "Large FCC ({}  atoms): {} pairs in {:.3}s",
        n_atoms,
        nl.len(),
        elapsed.as_secs_f64()
    );
}

#[test]
fn test_large_bcc_scaling_1000_atoms() {
    // 8x8x8 BCC supercell = 1024 atoms
    let large_bcc = make_bcc_large(2.87, Element::Fe, 8);
    let n_atoms = large_bcc.num_sites();
    assert!(
        n_atoms >= 1000,
        "Expected at least 1000 atoms, got {}",
        n_atoms
    );

    let config = NeighborListConfig {
        cutoff: 2.6, // First shell
        ..Default::default()
    };

    let start = Instant::now();
    let nl = build_neighbor_list(&large_bcc, &config);
    let elapsed = start.elapsed();

    // Performance
    assert!(
        elapsed.as_secs_f64() < 2.0,
        "Took {:.2}s for {} atoms",
        elapsed.as_secs_f64(),
        n_atoms
    );

    // Correctness: each atom should have 8 neighbors
    let mut counts = vec![0usize; n_atoms];
    for &center in &nl.center_indices {
        counts[center] += 1;
    }

    let all_8 = counts.iter().all(|&c| c == 8);
    assert!(all_8, "All BCC atoms should have CN=8");

    println!(
        "Large BCC ({} atoms): {} pairs in {:.3}s",
        n_atoms,
        nl.len(),
        elapsed.as_secs_f64()
    );
}

#[test]
fn test_large_nacl_scaling() {
    // 5x5x5 NaCl supercell = 1000 atoms
    let large_nacl = make_nacl_supercell(5.64, 5);
    let n_atoms = large_nacl.num_sites();
    assert!(
        n_atoms >= 1000,
        "Expected at least 1000 atoms, got {}",
        n_atoms
    );

    let config = NeighborListConfig {
        cutoff: 3.5, // First shell Na-Cl distance
        ..Default::default()
    };

    let start = Instant::now();
    let nl = build_neighbor_list(&large_nacl, &config);
    let elapsed = start.elapsed();

    // Performance
    assert!(
        elapsed.as_secs_f64() < 2.0,
        "Took {:.2}s for {} atoms",
        elapsed.as_secs_f64(),
        n_atoms
    );

    // Correctness: each atom should have 6 neighbors (rocksalt CN=6)
    let mut counts = vec![0usize; n_atoms];
    for &center in &nl.center_indices {
        counts[center] += 1;
    }

    let all_6 = counts.iter().all(|&c| c == 6);
    assert!(all_6, "All NaCl atoms should have CN=6");

    println!(
        "Large NaCl ({} atoms): {} pairs in {:.3}s",
        n_atoms,
        nl.len(),
        elapsed.as_secs_f64()
    );
}

// === Cutoff Comparison Tests ===

#[test]
fn test_cutoff_comparison_fcc() {
    // Compare neighbor counts at different cutoffs for FCC Cu
    let fcc = make_fcc(3.61, Element::Cu);

    // Expected coordination numbers for FCC at different shells
    // First shell: ~2.55 Å, CN=12
    // Second shell: ~3.61 Å, CN=6
    // Third shell: ~4.42 Å, CN=24

    let cutoffs = [2.5, 3.0, 4.0, 5.0];
    let mut prev_count = 0;

    for cutoff in cutoffs {
        let config = NeighborListConfig {
            cutoff,
            ..Default::default()
        };

        let nl = build_neighbor_list(&fcc, &config);
        let total_neighbors = nl.len() / 4; // per atom average

        // Neighbor count should be non-decreasing with cutoff
        assert!(
            total_neighbors >= prev_count,
            "Cutoff {}: {} neighbors < previous {}",
            cutoff,
            total_neighbors,
            prev_count
        );

        prev_count = total_neighbors;
        println!(
            "FCC Cu cutoff {:.1} Å: {} neighbors/atom",
            cutoff, total_neighbors
        );
    }
}

#[test]
fn test_cutoff_comparison_bcc() {
    // Compare neighbor counts at different cutoffs for BCC Fe
    let bcc = make_bcc(2.87, Element::Fe);

    // BCC: first shell ~2.48 Å (CN=8), second shell ~2.87 Å (CN=6)

    let cutoffs = [2.5, 3.0, 4.0, 5.0];
    let mut prev_count = 0;

    for cutoff in cutoffs {
        let config = NeighborListConfig {
            cutoff,
            ..Default::default()
        };

        let nl = build_neighbor_list(&bcc, &config);
        let total_neighbors = nl.len() / 2; // per atom average

        assert!(
            total_neighbors >= prev_count,
            "Cutoff {}: {} neighbors < previous {}",
            cutoff,
            total_neighbors,
            prev_count
        );

        prev_count = total_neighbors;
        println!(
            "BCC Fe cutoff {:.1} Å: {} neighbors/atom",
            cutoff, total_neighbors
        );
    }
}

#[test]
fn test_cutoff_comparison_nacl() {
    // Compare neighbor counts at different cutoffs for NaCl
    let nacl = make_nacl(5.64);

    // NaCl: Na-Cl first shell ~2.82 Å (CN=6)

    let cutoffs = [2.5, 3.0, 4.0, 5.0];
    let mut prev_count = 0;

    for cutoff in cutoffs {
        let config = NeighborListConfig {
            cutoff,
            ..Default::default()
        };

        let nl = build_neighbor_list(&nacl, &config);
        let total_neighbors = nl.len() / 8; // per atom average

        assert!(
            total_neighbors >= prev_count,
            "Cutoff {}: {} neighbors < previous {}",
            cutoff,
            total_neighbors,
            prev_count
        );

        prev_count = total_neighbors;
        println!(
            "NaCl cutoff {:.1} Å: {} neighbors/atom",
            cutoff, total_neighbors
        );
    }
}

#[test]
fn test_distance_distribution_consistency() {
    // Verify that distance distributions are consistent across cutoffs
    let fcc = make_fcc(3.61, Element::Cu);
    let _nn_dist = 3.61 / 2.0_f64.sqrt(); // First shell distance

    // Get neighbors at cutoff 3.0
    let config_small = NeighborListConfig {
        cutoff: 3.0,
        ..Default::default()
    };
    let nl_small = build_neighbor_list(&fcc, &config_small);

    // Get neighbors at cutoff 5.0
    let config_large = NeighborListConfig {
        cutoff: 5.0,
        ..Default::default()
    };
    let nl_large = build_neighbor_list(&fcc, &config_large);

    // All distances from small cutoff should appear in large cutoff
    let small_dists: std::collections::HashSet<i64> = nl_small
        .distances
        .iter()
        .map(|d| (d * 1000.0).round() as i64) // round to 0.001 Å
        .collect();

    let large_dists: std::collections::HashSet<i64> = nl_large
        .distances
        .iter()
        .map(|d| (d * 1000.0).round() as i64)
        .collect();

    for dist in &small_dists {
        assert!(
            large_dists.contains(dist),
            "Distance {} from small cutoff not found in large cutoff",
            *dist as f64 / 1000.0
        );
    }
}

#[test]
fn test_supercell_consistency() {
    // Neighbor counts should be the same for unit cell and supercell
    let unit_cell = make_fcc(3.61, Element::Cu);
    let supercell = make_fcc_large(3.61, Element::Cu, 3);

    let config = NeighborListConfig {
        cutoff: 3.0,
        ..Default::default()
    };

    let nl_unit = build_neighbor_list(&unit_cell, &config);
    let nl_super = build_neighbor_list(&supercell, &config);

    // Count per atom
    let unit_per_atom = nl_unit.len() / unit_cell.num_sites();
    let super_per_atom = nl_super.len() / supercell.num_sites();

    assert_eq!(
        unit_per_atom, super_per_atom,
        "Unit cell ({}) and supercell ({}) should have same CN",
        unit_per_atom, super_per_atom
    );
}
